glom/core.py

```diff
@@ -2531,3 +2537,38 @@ def FILL(target, spec, scope):
     if callable(spec):
         return spec(target)
     return spec
+
+class _ArgValuator(object):
+    def __init__(self):
+        self.cache = {}
+
+    def mode(self, target, spec, scope):
+        """
+        similar to FILL, but without function calling;
+        useful for default, scope assignment, call/invoke, etc
+        """
+        recur = lambda val: scope[glom](target, val, scope)
+        result = spec
+        if type(spec) in (list, dict):  # can contain themselves
+            if id(spec) in self.cache:
+                return self.cache[id(spec)]
+            result = self.cache[id(spec)] = type(spec)()
+            if type(spec) is dict:
+                result.update({recur(key): recur(val) for key, val in spec.items()})
+            else:
+                result.extend([recur(val) for val in spec])
+        if type(spec) in (tuple, set, frozenset):  # cannot contain themselves
+            result = type(spec)([recur(val) for val in spec])
+        return result
+
+
+def arg_val(target, arg, scope):
```

Comment:
should `arg_val` be more documented and public? Just thinking that if others want to develop specs, then they'll conventionally want to use this for arguments?

Reply:
yea, that's a good point -- maybe arg_val is the right thing to make public and throw a big comment on

as I was playing around with the compilation I thought maybe it could work kind of like the re module:

arg_val = re.match

arv_compile = re.compile

so, arg_val and arg_compile would be the public entry points

in which case, maybe we should take a minute to think if arg_val is the right name: should it be arg_eval?  arg_glom?

Reply:
sorry, to round out the re parallel:

like re, you can do arg_val(arg_compile(arg)) or arg_val(arg), and they'll both give the same result

like re, its up to the owning spec to figure out where it wants to store the result of arg_compile